packages/flutter/lib/src/widgets/image_filter.dart

```diff
@@ -0,0 +1,60 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+import 'framework.dart';
+
+/// Applies an [ImageFilter] to its child.
+@immutable
+class ImageFiltered extends SingleChildRenderObjectWidget {
+  /// Creates a widget that applies an [ImageFilter] to its child.
+  ///
+  /// The [imageFilter] must not be null.
```

Comment:
I wonder if this is too restrictive. It means that if you want to change filtering at runtime you have to change your widget hierarchy, which can be quite ceremonious in the code. Maybe instead use the strategy used by `Opacity` and `Transform`? They first check if the widget has any effect at all, and if it doesn't simply does not apply the effect. This does not require users to dynamically change their widget trees, but only values passed to the `Opacity`/`Transform` widgets.

Reply:
I was thinking that as well, but it seems a common restriction in the code.

There are even things like ClipRect asserts that the behavior isn't "None"...? Then why do we define the value?

BDF also asserts a non-null filter as well.

Is there a reason for not allowing these widgets to have a "nop" mode in them?

@Hixie ?

Reply:
`ClipBehavior.none` is allowed by most widgets, for `ClipRect` specifically it doesn't make sense because the whole point is to clip.

Asserting non-null here is fine IMHO.